### ISSUE TITLE
Download attempt while offline - unexpected behavior, plus null data retrieved - that's an error

### ISSUE THREAD

**Issue by Twv** [dropbox/dropbox-sdk-obj-c#69]
I would like to ask what the expected behavior is when attempting download. E.g. below example:
```objectivec
DBDownloadDataTask *dTask = [client.filesRoutes downloadData:file];
[dTask response:^(DBFILESFileMetadata* _Nullable result, DBFILESDownloadError* _Nullable routeError, DBRequestError * _Nullable error, NSData * _Nonnull data) {
     ....handle response....
}];
```
I am testing a scenario where device has a working connection at the beginning and retrieves a list of files. But after a while, the connection drops for some reason. (I turn on airplane mode on the debug device).

An attempt to download a file is made. What is the expected result when attempting a download without connection (calling code above)?

What happens is:
- The task waits until the connection is re-established. No timeout or error occurs, no matter how long I let the task wait.
- When connection is re-established (I turn off the airplane mode), response arrives... but NSData * _Nonnull data parameter contains no content. routeError, error are both nil also.

I expected the response to be called immediately and an error (no connection) would be reported.
Such behavior is for example observed when listing files:
```objectivec
NSNumber *includeMediaInfo = [NSNumber numberWithBool:YES];
[[client.filesRoutes listFolder:cloudPath recursive:nil includeMediaInfo:includeMediaInfo includeDeleted:nil includeHasExplicitSharedMembers:nil]
   response:^(DBFILESListFolderResult * _Nullable result, DBFILESListFolderError * _Nullable routeError, DBRequestError * _Nullable error) {
     ....handle response....
}];
```
If the above code is attempted when Airplane mode is turned on, response is triggered immediately and an error is reported.

I tried to debug the code - when list is called with airplane mode on, code steps into Shared/Handwritten/Networking/DBDelegate.m:
```objectivec
- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
```
But when the same scenario is repeated for download, this function is never called.

I don't know if this is by design or not. But I'd kind of need to know how to handle such requests - I don't want the UI to keep waiting on a download that will never arrive and I'd hate to have to implement a parallel connectivity/timeout checking.

TLDR - what is the expected result when calling download with airplane mode on?
After the airplane mode is turned off, the response is triggered, but no data is received.

Edit: I'm using api version from today's github master (54b1f8f).

BR, Jan.

**Comment by scobbe** [comment]
@Twv: Thanks for reaching out. Yes, this is a problem I noticed when writing the SDK. I wasn't able to come up with a good solution.

The problem is described well [here](http://stackoverflow.com/questions/23288780/nsurlsessiontask-never-calls-back-after-timeout-when-using-background-configurat). Basically the problem only arises when using a background `NSURLSession`.

I can spend some time tomorrow revisiting the issue and see if I can come up with a better solution.

**Comment by Twv** [comment]
Thank you for your response.
I've read the link - uff. If I understand correctly - a network falling out in the middle of the download would not trigger any event. The download would just stall indefinitely - well up to a week :).

This means I need to make a wrapper around all dropbox calls, monitor the task progress events and monitor network reachability and then combine both into network issue handling... 

Do you have any suggestion on how to implement calls such, that and offline network would consistently trigger an error and end the task?

Also - I've done some extra testing on the nil data being returned from the task without any error given - it occurs sporadically. It feels a lot like an async issue - it doesn't happen a lot - but it does.

BR, Jan.

**Comment by scobbe** [comment]
Thanks for your reply, Jan.

> Do you have any suggestion on how to implement calls such, that an offline network would consistently trigger an error and end the task?

I'll spend some time later today and see if I can implement logic in the SDK to do this. Really this should be handled by the SDK and should not be required for every client that integrates with the SDK. Let me get back to you tomorrow or the day after with a progress update.

> Also - I've done some extra testing on the nil data being returned from the task without any error given - it occurs sporadically. It feels a lot like an async issue - it doesn't happen a lot - but it does.

Sorry that you're running into issues here. Would you mind posting some steps to reproduce your test setup? I haven't seen this issue before, so it would be helpful to know how I can try and replicate it.

Thank you!

**Comment by xaphod** [comment]
Hi @scobbe,
Just a quick remark -- per that SO thread you linked, wouldn't the easiest workaround be to change the backgroundSession's `timeoutIntervalForResource` to less than a week (in the NSURLSessionConfiguration object when you create the session) ?
I didn't check but probably the SDK already allows us to specify our own background session (rather than have the SDK use the default)... so the app using the SDK could simply define its own `timeoutIntervalForResource` as it needs in a backgroundSession and pass that in...

**Comment by scobbe** [comment]
@xaphod: Thanks for your thoughts. Unless I haven't properly set it up, I think the SDK does that already [here](https://github.com/dropbox/dropbox-sdk-obj-c/blob/master/Source/ObjectiveDropboxOfficial/Shared/Handwritten/Networking/DBTransportClient.m#L68). If I remember, though, even setting this did not address the issue of an interrupted connection. But let me check and test it again.

**Comment by xaphod** [comment]
@scobbe Per that SO thread, there are 2 timeouts. The `timeoutIntervalForRequest` will never take effect for an interrupted connection. But in my experience, `timeoutIntervalForResource ` is reliable.

Also please note that using airplane mode is not a good way to test: things behave very differently with airplane mode vs a regular internet drop.
To simulate a regular internet drop, I suggest you get a wifi router on your desk (ie. do this at home with your home wifi), turn OFF cellular data entirely, start a dataTask or downloadTask in the background, then turn OFF the router. In other words don't change settings on the phone to simulate the failure, kill something else. This works a lot better in my testing.

**Comment by scobbe** [comment]
@xaphod: Ah actually just re-read the thread – looks like I'm setting `timeoutIntervalForRequest` and not `timeoutIntervalForResource`. Let me update that and try now. Thanks for the ping!

**Comment by xaphod** [comment]
haha when comments collide....

**Comment by scobbe** [comment]
Haha thanks for following up with this! I'll try your suggestions now 👍

**Comment by xaphod** [comment]
@scobbe Cool. I'm actually just in the process of migrating my iOS app "Grab for Dropbox" to API v2. It does **exactly** what this thread here discusses: its whole raison d'être is to help the user download (offline) huuuuge amounts of photos and videos from dropbox, recursively. And it definitely has to be super tolerant against connectivity-loss & battery-dropout situations.

https://itunes.apple.com/us/app/grab-for-dropbox-shared-link/id1032995614?mt=8

**Comment by scobbe** [comment]
@xaphod: Nice! Be sure to let us know then if you run into any performance issues using the SDK during your migration.

@Twv: Thanks to @xaphod 's suggestion, this should now be fixed in the latest version of the SDK (v2.0.3). Please pull the latest version and see if you still experience your problem.

I tested it out and the SDK now seems to handle bad network connections well. I set the timeout period to 60 seconds, so within a minute of losing connection, the response handler will be executed.

That said, I would advise against blocking the UI thread waiting for the download to finish. Be sure to execute `downloadUrl` or whichever download method you're using with an `NSOperationQueue` that's not the main queue, to avoid blocking things for the user.

**Comment by Twv** [comment]
Thank you for your effort.
I will give it a spin asap.

Meanwhile, I/d just like to comment on the NSURLSession.
It seems to me, that the issue here is not really with the bug and not-bug. It is about the two different use-cases being squashed into a single api call.

I see two use-cases:
- first is the interactive use, where a user clicks on a file, is presented a download in progress dialog (as is my case - I'm not fetching on the UI thread), and should be notified if something failed during the download. If all is well, user gets the file opened on her device after the download finishes.
- second is the background download - where a user would select N files (or queue a large file for download), press a download button and then switch application. User doesn't care about the immediate feedback here - she just wants to get the files at some point in the future (this seems to be your supported use case in the API).

Using the background URL session supports the second use-case, but not the first one.

Point is - I now understand the issue. This is more an inconsistent behaviour when using different api calls (feedback differs when fetching directory list/requesting a file delete and on the other hand downloading/uploading a file) - maybe there should be a way to instruct API which kind of session (background or not) to use for a call.

BR, Jan.

**Comment by scobbe** [comment]
@Twv: Thanks again for bringing these issues to our attention. So after reading your post and doing a little more research, I think we're going to keep the background session timeout behavior as it was before, i.e., no timeout. I think that will fit the most common user use-case the best.

That said, to accommodate the "interactive use" case (as you put it), I've gone ahead and exposed a "force foreground session" initializer in the transport client, which makes the SDK use a foreground session for all actions, which may be more in line with what you want.

To use this new feature, in your app delegate, you would want to use the following code:

```
DBTransportClient *transportClient = [[DBTransportClient alloc] initWithForceForegroundSession];
[DropboxClientsManager setupWithAppKey:@"<APP_KEY>" transportClient:transportClient];
```

Also, I think I figured out the issue with the bug where no error data was being returned for the download to data method. That should be fixed now with the latest version (2.0.5), so please let us know if you see the issue again.

In the meantime, please let us know if this new feature sufficiently addresses your use case.

**Comment by Twv** [comment]
Thank you for your effort.
The change is good. I can now use the download with a foreground session.
